File: codeninja/01 pair sum/sol2.cpp
#include <bits/stdc++.h>
#include <iostream>
#include <map>

using namespace std;

vector<vector<int>> pairSum(vector<int> &arr, int s){
   int n = arr.size();

    // Map to store frequency of elements.
    map<int, int> hash;

    // This will store the result.
    vector<vector<int>> ans;

    for (int ele : arr) {
        int count = hash[s - ele];

        vector<int> pair(2);
        pair[0] = ele;
        pair[1] = s - ele;

        while (count--)
            ans.push_back(pair);

        hash[ele] += 1;
    }

    // This will store the final result.
    vector<vector<int>> res(ans.size(), vector<int>(2, 0));

    // Sorting the 'ans' arrays element.
    for (int i = 0; i < ans.size(); i++) {
        int a = ans[i][0], b = ans[i][1];
        res[i][0] = min(a, b);
        res[i][1] = max(a, b);
    }

    // Finally sorting each pair in 'res'.
    sort(res.begin(), res.end());
    return res;
}

int main() {

   vector<int> arr = {10,20,30,40,50};
   vector<vector<int>> res = pairSum(arr,50);

   for(vector<int> v : res){
       for(int i : v)
            cout << i << " ";
        cout << endl;
   }

   return 0;
}

File: coding/01 pair sum/sol2.java
import java.util.*;

public class HelloWorld{

    public static int[][] pairSum(int arr[], int s) {
        int n = arr.length;

        // HashMap to store frequency of elements.
        HashMap<Integer, Integer> map = new HashMap();

        // This will store the result.
        List<int[]> ans = new ArrayList();
        for (int ele : arr) {
            int count = map.getOrDefault(s - ele, 0);

            int[] pair = new int[2];
            pair[0] = ele;
            pair[1] = s - ele;

            while (count-- > 0) {
                ans.add(pair);
            }

            map.put(ele, map.getOrDefault(ele, 0) + 1);
        }

        // This will to store final result.
        int res[][] = new int[ans.size()][2];

        // Sorting the 'ans' arrays element.
        for (int i = 0; i < ans.size(); i++) {
            int a = ans.get(i)[0], b = ans.get(i)[1];
            res[i][0] = Math.min(a, b);
            res[i][1] = Math.max(a, b);
        }

        // Finally sorting each pair in 'res'.
        Arrays.sort(res, new Comparator<int[]>() {
            public int compare(int[] a, int[] b) {
                if (a[0] == b[0]) {
                    return a[1] - b[1];
                }
                return a[0] - b[0];
            }
        });

        return res;
    }

    public static void main(String []args){
        int[] arr;

        arr = new int[5];
        arr[0] = 10;
        arr[1] = 20;
        arr[2] = 30;
        arr[3] = 40;
        arr[4] = 50;

        int[][] res = pairSum(arr, 50);

        for (int i = 0; i < res.length; i++){
            for (int j = 0; j < res[i].length; j++)
                System.out.print(res[i][j] + " ");
            System.out.println(" ");
        }

        return;
     }
}